pennylane_lightning/src/algorithms/AdjointDiff.hpp

```diff
@@ -722,6 +499,114 @@ template <class T = double> class AdjointJacobian {
             }
         }
     }
-}; // class AdjointJacobian
 
+    /**
+     * @brief Calculates the Jacobian for the statevector for the selected set
+     * of parametric gates.
+     *
+     * For the statevector data associated with `psi` of length `num_elements`,
+     * we make internal copies to a `%StateVectorManaged<T>` object, with one
+     * per required observable. The `operations` will be applied to the internal
+     * statevector copies, with the operation indices participating in the
+     * gradient calculations given in `trainableParams`, and the overall number
+     * of parameters for the gradient calculation provided within `num_params`.
+     * The resulting row-major ordered `jac` matrix representation will be of
+     * size `jd.getSizeStateVec() * jd.getObservables().size()`. OpenMP is used
+     * to enable independent operations to be offloaded to threads.
+     *
+     * @param jac Preallocated vector for Jacobian data results.
+     * @param jd JacobianData represents the QuantumTape to differentiate
+     * @param apply_operations Indicate whether to apply operations to tape.psi
+     * prior to calculation.
+     */
+    void adjointJacobianJD(std::vector<T> &jac, const JacobianData<T> &jd,
```

Comment:
Rather the reimplement the internal structure, is it possible for this method to offload to the `adjointJacobian` with the requested datatypes from `jd`? 

That way we can avoid having the same algo implemented twice, and makes testing earier. What do you think?

Reply:
Thanks @mlxd for your review! They are now two `adjointJacobianJD` and `adjointJacobian`... I only kept the other one for testing and comparing. If you are +1 with using a data-structure such as `JacobianData`, I'll replace `adjointJacobian` by `adjointJacobianJD`. The result of this PR must not include two implementations for adjoint-jacobian :slightly_smiling_face: